### ISSUE TITLE
bug in solver for band matrices

### ISSUE THREAD

**Issue by conradsnicta** [conradsnicta/armadillo-code#76]
For [band matrices](https://en.wikipedia.org/wiki/Band_matrix), the [solve()](http://arma.sourceforge.net/docs.html#solve) function currently misdetects the number of upper diagonals.  Bug is in `band_helper::is_band()`.

As a temporary workaround, it's possible to use the `solve_opts::no_band` option. Example:

    solve(x, A, b, solve_opts::no_band)

@eddelbuettel

**Comment by conradsnicta** [comment]
fixed in:
https://github.com/conradsnicta/armadillo-code/commit/5ed55c23522614649af2ff8cee0f50d01ebcef23 (8.500.x branch)
https://github.com/conradsnicta/armadillo-code/commit/9480ef781b7e4d6671509c4fac74a7ee31ef96a1 (unstable branch)

**Comment by eddelbuettel** [comment]
Thanks for the fix, I updated my source repo.

**Comment by eddelbuettel** [comment]
And now pushed to the drat in https://github.com/RcppCore/drat/commit/39df94ab59f0f270fecb6f794467857535c57375.

**Comment by conradsnicta** [comment]
@eddelbuettel @kthohr - I've extended the band solver to use specialised handling of tri-diagonal matrices, resulting in notable speedups for that particular use case. The following lapack functions are employed: `sgtsv`, `sgtsvx`, `dgtsv`, `dgtsvx`, `cgtsv`, `cgtsvx`, `zgtsv`, `zgtsvx`.

The `dgtsv`and `dgtsvx` functions (for double-precision matrices) appear to be present in the borked version of lapack that R insists on shipping.  As usual, the complex number versions (eg. `zgtsv`, `zgtsvx`) are not present in borked lapack. Their use in Armadillo is guarded via ARMA_CRIPPLED_LAPACK.

**Comment by eddelbuettel** [comment]
Nice.  

(And once again, for clarity, those "borked" LAPACK are a _fallback_ used when nothing better is available.  That seems to be the default on CentOS/RHEL but I ensure that Debian/Ubuntu builds use an external LAPACK/BLAS which is complete.  People who think they know better and build from source get what they deserve, the cripped fallback.  Then again, in R land we do not see too many banded or complex matrices so ...)

**Comment by kthohr** [comment]
Indeed, `zgtsv` and `zgtsvx` appear to be missing. I can run a check on the M$ side with reduced Lapack, if needed.

**Comment by conradsnicta** [comment]
@kthohr - that would be very useful; the more testing the better

**Comment by kthohr** [comment]
I dropped the `unstable` branch into an RcppArmadillo build, on a Windows machine with reduced Lapack. The existing Rlapack unit tests passed without any problems. 

In order to trigger `solve_tridiag_refine`, I noticed there's a requirement on size of `A`:
https://github.com/conradsnicta/armadillo-code/blob/34a1f79440e7bb3f755cf68b8ca3073685a9d9f2/include/armadillo_bits/band_helper.hpp#L38

In `glue_solve_gen`, `N_min` is 32. Do you think this value is likely to change in the near future? It would need to be hardcoded into a unit test on the RcppArmadillo side, as the size of the Rlapack test matrices are much smaller than this.

**Comment by kthohr** [comment]
I've added an additional test for the tridiagonal case (to trigger `zgtsvx`).  Might be useful when @eddelbuettel updates the base Armadillo code.
